### PR TITLE
Fix SFT dataloader checkpoint bug (#745)

### PR DESCRIPTION
**Pull request by baonudesifeizhai**
- Add get_state() and set_state() methods to SFTDataset and PackingDataset
- Modify CheckpointManager to save/load dataloader state
- Update SFT training loop to handle dataloader state properly
- Fixes issue where resuming training would see redundant samples

 https://github.com/PrimeIntellect-ai/prime-rl/issues/745

### CODE REVIEW

File: src/prime_rl/trainer/ckpt.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ def _save_to_path(
         optimizers: list[Optimizer],
         scheduler: LRScheduler,
         progress: Progress,
+        dataloader_state: dict | None = None,
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/ckpt.py]: can we not use the same pattern as for the other ckpt objects and pas the object to the function and convert to state dict inside?

Comment (mikasenghaas) [reply src/prime_rl/trainer/ckpt.py]: also, why is it optional?

File: src/prime_rl/trainer/ckpt.py [hunk 2]
```diff
@@ -102,12 +107,28 @@ def _load_from_path(
...
+        for optimizer, optimizer_state in zip(optimizers, state["optimizers"]):
+            optimizer.load_state_dict(optimizer_state)
+        scheduler.load_state_dict(state["scheduler"])
+
+        # Load progress
+        for key, value in asdict(state["progress"]).items():
+            setattr(progress, key, value)
+        
+        # Return dataloader state if available
+        return state.get("dataloader_state")
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/ckpt.py]: again, i would like this to be in place to follow the pattern of the checkpoint class. this means, that we would pass the dataloader to the `load` function and update its state in-place. the `load` function should keep returning `None`

File: src/prime_rl/trainer/sft/data.py [hunk 3]
```diff
@@ -82,18 +82,27 @@ def __init__(self, tokenizer: PreTrainedTokenizer, config: DataConfig):
...
+        self._epoch = -1
 
     def __iter__(self) -> Iterator[Sample]:
         """
         Apply chat template and tokenize a single example in prompt + completion format (https://github.com/huggingface/trl/blob/de27d612b026526ba39b88eee348994d7636e033/trl/trainer/sft_trainer.py#L661)
         """
-        counter, epoch = -1, -1
+        # Use instance state for checkpointing
+        counter = self._counter
+        epoch = self._epoch
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/sft/data.py]: can we not just use `self._counter` and `self._epoch`? seems like they are fully redundant with `counter` and `epoch`. if so, let's ditch the local variables and make the class attributes public, i.e. `self.counter` and `self.epoch`

File: src/prime_rl/trainer/sft/data.py [hunk 4]
```diff
@@ -136,6 +145,21 @@ def __iter__(self) -> Iterator[Sample]:
...
+            "epoch": self._epoch,
+            "data_rank": self.data_rank,
+            "data_world_size": self.data_world_size,
+        }
+    
+    def set_state(self, state: dict):
+        """Set the state from checkpoint."""
+        self._counter = state["counter"]
+        self._epoch = state["epoch"]
+        # Note: data_rank and data_world_size are not restored as they depend on world config
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/sft/data.py]: hm, what does this mean if we wanted to resume training with a different number of gpus? in general, it seems weird that the get state and set state function do not operate on the same attributes, seems off..

File: src/prime_rl/trainer/sft/data.py [hunk 5]
```diff
@@ -165,6 +189,21 @@ def __iter__(self) -> Iterator[Sample]:
                         packed_samples[key] = value[: self.seq_len]
                 yield packed_samples
                 packed_samples, seq_len = defaultdict(list), 0
+    
+    def get_state(self) -> dict:
+        """Get the current state for checkpointing."""
+        # PackingDataset doesn't maintain state, but its underlying dataset does
+        if hasattr(self.dataset, 'get_state'):
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/sft/data.py]: when would this ever be false?

File: src/prime_rl/trainer/sft/train.py [hunk 6]
```diff
@@ -75,19 +75,31 @@ def train(config: SFTTrainerConfig):
...
         logger.info(f"Resuming training from checkpoint step `{config.ckpt.resume_step}`")
-        ckpt_manager.load(model, [optimizer], scheduler, progress, step=config.ckpt.resume_step)
+        dataloader_state = ckpt_manager.load(model, [optimizer], scheduler, progress, step=config.ckpt.resume_step)
+        
+        # Restore dataloader state if available
+        if dataloader_state is not None and hasattr(dataloader_obj, 'set_state'):
+            logger.info("Restoring dataloader state from checkpoint")
+            dataloader_obj.set_state(dataloader_state)
+        else:
+            logger.warning("No dataloader state found in checkpoint, starting from beginning")
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/sft/train.py]: if we make the load in-place this should just be `ckpt_manager.load(model, [optimizer], scheduler, dataloader, progress, step=config.ckpt.resume_step)` which is much cleaner imo

File: src/prime_rl/trainer/sft/train.py [hunk 7]
```diff
@@ -105,7 +117,13 @@ def train(config: SFTTrainerConfig):
...
             logger.info(f"Saving checkpoint at step {progress.step}")
             save_ckpt_start_time = time.time()
-            ckpt_manager.save(model, [optimizer], scheduler, progress, step=progress.step)
+            
+            # Get dataloader state for checkpointing
+            dataloader_state = None
+            if hasattr(dataloader_obj, 'get_state'):
+                dataloader_state = dataloader_obj.get_state()
+            
+            ckpt_manager.save(model, [optimizer], scheduler, progress, step=progress.step, dataloader_state=dataloader_state)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/sft/train.py]: same here, would just be `ckpt_manager.save(model, [optimizer], scheduler, dataloader, progress, step=progress.step)`

File: src/prime_rl/trainer/sft/train.py [hunk 8]
```diff
@@ -105,7 +117,13 @@ def train(config: SFTTrainerConfig):
         ):
             logger.info(f"Saving checkpoint at step {progress.step}")
             save_ckpt_start_time = time.time()
-            ckpt_manager.save(model, [optimizer], scheduler, progress, step=progress.step)
+            
+            # Get dataloader state for checkpointing
+            dataloader_state = None
+            if hasattr(dataloader_obj, 'get_state'):
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/sft/train.py]: when would this be false?

**Review by mikasenghaas — COMMENTED** [review]
let's give the dataloader a `load_state_dict` function for loading in-place and `state_dict` for getting the current state so that we mirror that torch api for checkpointing. this way we have minimal changes to the orchestrator checkpointing logic.

also, we should think about how resuming a sft training would look on different number of gpus. is a epoch-aware counter into the dataset enough?

**Comment by baonudesifeizhai** [comment]
new pr solve this problem...i will close

**Comment by mikasenghaas** [comment]
> new pr solve this problem...i will close

@baonudesifeizhai wait, what? which one?

**Comment by baonudesifeizhai** [comment]
> > new pr solve this problem...i will close
> 
> @baonudesifeizhai wait, what? which one?

https://github.com/PrimeIntellect-ai/prime-rl/pull/833  this one ...?anyway  i never thought that way ....my bad

**Comment by baonudesifeizhai** [comment]
> > new pr solve this problem...i will close
> 
> @baonudesifeizhai wait, what? which one?

https://github.com/PrimeIntellect-ai/prime-rl/pull/833   this one

**Comment by mikasenghaas** [comment]
@baonudesifeizhai it doesn't tho? #833 changes how FSDP distributed checkpoints are written to allow restarting on dynamic world sizes. We still need to properly checkpoint the SFT dataloader, though you probably want to rebase onto main to checkpoint using the updated trainer's `CheckpointManager` which now uses the stateful `AppState`. let me know if you want to pursue it, we will need this on in pretty soon for a larger SFT run we have planned, for which we want proper checkpointing of the dataloader

**Comment by baonudesifeizhai** [comment]
> @baonudesifeizhai it doesn't tho? #833 changes how FSDP distributed checkpoints are written to allow restarting on dynamic world sizes. We still need to properly checkpoint the SFT dataloader, though you probably want to rebase onto main to checkpoint using the updated trainer's `CheckpointManager` which now uses the stateful `AppState`. let me know if you want to pursue it, we will need this on in pretty soon for a larger SFT run we have planned, for which we want proper checkpointing of the dataloader
 

> @baonudesifeizhai it doesn't tho? #833 changes how FSDP distributed checkpoints are written to allow restarting on dynamic world sizes. We still need to properly checkpoint the SFT dataloader, though you probably want to rebase onto main to checkpoint using the updated trainer's `CheckpointManager` which now uses the stateful `AppState`. let me know if you want to pursue it, we will need this on in pretty soon for a larger SFT run we have planned, for which we want proper checkpointing of the dataloader

i will find a way, ethier rebase or rewrite the code base on #833 ....

**Comment by baonudesifeizhai** [comment]
> @baonudesifeizhai it doesn't tho? #833 changes how FSDP distributed checkpoints are written to allow restarting on dynamic world sizes. We still need to properly checkpoint the SFT dataloader, though you probably want to rebase onto main to checkpoint using the updated trainer's `CheckpointManager` which now uses the stateful `AppState`. let me know if you want to pursue it, we will need this on in pretty soon for a larger SFT run we have planned, for which we want proper checkpointing of the dataloader

i  use   (DCP)  save dataloader status...and test on your platform seems work,,,it really heart... could you give me some credit on prime-rl...my gmail: baonudesifeizhai@gmail.com
thanks bro!

**Comment by baonudesifeizhai** [comment]
> @baonudesifeizhai it doesn't tho? #833 changes how FSDP distributed checkpoints are written to allow restarting on dynamic world sizes. We still need to properly checkpoint the SFT dataloader, though you probably want to rebase onto main to checkpoint using the updated trainer's `CheckpointManager` which now uses the stateful `AppState`. let me know if you want to pursue it, we will need this on in pretty soon for a larger SFT run we have planned, for which we want proper checkpointing of the dataloader

i use (DCP) save dataloader status...and test on your platform seems work..

**Comment by mikasenghaas** [comment]
nice, will check this tmr!

**Comment by mikasenghaas** [comment]
Closing in favor if #849, you will get your contribution from that PR:)
